lti_consumer/plugin/views.py

```diff
@@ -420,3 +424,106 @@ def scores(self, request, *args, **kwargs):  # pylint: disable=unused-argument
             status=status.HTTP_201_CREATED,
             headers=headers
         )
+
+
+class LtiNrpsContextMembershipViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    LTI NRPS Context Membership Service endpoint.
+
+    See full documentation at:
+    http://imsglobal.org/spec/lti-nrps/v2p0
+    """
+
+    # Custom permission classes for LTI APIs
+    authentication_classes = [Lti1p3ApiAuthentication]
+    permission_classes = [LtiNrpsContextMembershipsPermissions]
+
+    # Renderer classes to accept LTI NRPS content types
+    renderer_classes = [
+        MembershipResultRenderer,
+    ]
+
+    def get_profile_images(self, users):
+        """
+        Given a list of users, returns a dict mapping of corresponding profile images.
+        """
+        # get profile images
+        profile_images = {}
+        for user in users:
+            # if user's image exists turn it to an absolute url
+            if user.profile.has_profile_image:
+                profile_images[user.id] = compat.get_user_profile_image(user, self.request)['small']
+        return profile_images
+
+    def preprocess(self, data):
+        """
+        Preprocess the output of `get_membership` method. It now does followings -
+            - Append external ids to the user
+            - Makes user profile picture link absolute
+        """
+        processed_data = []
+
+        # batch get or create external ids for all users
+        user_ids = data.keys()
+        users = get_user_model().objects.prefetch_related('profile').filter(id__in=user_ids)
+
+        # get external ids
+        external_ids = compat.batch_get_or_create_externalids(users)
+
+        # get profile images
+        profile_images = self.get_profile_images(users)
+
+        # import pdb; pdb.set_trace()
+        for userid, user_info in data.items():
+
+            # append external ids to user dictonary
+            user_info['external_id'] = external_ids[userid].external_user_id
+
+            if profile_images.get(userid):
+                user_info['picture'] = profile_images[userid]
+
+            processed_data.append(user_info)
+
+        return processed_data
+
+    def get_serializer_class(self):
+        """
+        Overrides ModelViewSet's `get_serializer_class` method.
+        Checks if PII fields can be exposed and returns appropiate serializer.
+        """
+        if expose_pii_fields(self.request.lti_configuration.location.course_key):
+            return LtiNrpsContextMembershipPIISerializer
+        else:
+            return LtiNrpsContextMembershipBasicSerializer
+
+    def list(self, *args, **kwargs):
+        """
+        Overrides default list method of ModelViewSet. Calls LMS `get_course_members`
+        API and returns result.
+        """
+
+        # get course key
+        course_key = self.request.lti_configuration.location.course_key
+
+        try:
+            data = compat.get_course_members(course_key)
```

Comment:
@shimulch How are you paginating these results? This still needs to be paginated here to respect the LTI spec parameters.

Reply:
@giovannicimolin, I thought we've decided to dump all data at once since we only going to support a certain limit on enrollment. So this API returns all data at once.

I think it is respecting LTI spec for [pagination](https://www.imsglobal.org/spec/lti-nrps/v2p0#limit-query-parameter), as we are not sending `rel="next"` in the header, it means there is no data. And the pagination spec seems optional.

Reply:
Makes sense, I've read the spec and checked that this is not a required parameter and that the LMS can ignore it.